test/Gen.hs

```diff
@@ -21,60 +21,72 @@ module Gen
 , Rec(..)
 , forall
   -- * Showing generated values
-, With(getWith)
-, showing
-, showingFn
 , atom
-, liftWith
-, liftWith2
-  -- * Pattern synonyms
-, pattern With
-, pattern Fn
-, pattern FnWith
+, Gen.label
+, infixL
+, infixR
+, addLabel
   -- * Re-exports
 , Gen
 , (===)
 , (/==)
-, choice
-, subterm
-, subtermM
-, subterm2
-, Arg
-, Vary
-, fn
-, apply
+, Gen.choice
+, Fn.Arg
+, Fn.Vary
+, Gen.fn
+, Fn.apply
 ) where
 
+import Control.Applicative
 import Control.Carrier.Pure
-import Data.Function (on)
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Writer
+import Data.Foldable (traverse_)
 import Data.Functor.Classes (showsUnaryWith)
 import Data.Proxy
+import qualified Data.Semigroup as S
+import qualified Data.Set as Set
+import Data.String (fromString)
 import GHC.Stack
 import GHC.TypeLits
-import Hedgehog
-import Hedgehog.Function hiding (R, S)
+import Hedgehog hiding (Gen)
+import qualified Hedgehog
+import qualified Hedgehog.Function as Fn
 import Hedgehog.Gen
 import Hedgehog.Range
 
 -- | A generator for computations, given a higher-order generator for effectful operations, & a generator for results.
 m
   :: forall m a
-  .  (Applicative m, Show a)
-  => (forall a . Show a => (forall a . Show a => Gen a -> Gen (With (m a))) -> Gen a -> Gen (With (m a))) -- ^ A higher-order generator producing operations using any effects in @m@.
-  -> Gen a                                                                                      -- ^ A generator for results.
-  -> Gen (With (m a))                                                                           -- ^ A generator producing computations, wrapped in 'With' for convenience.
+  .  Monad m
+  => (forall a . (forall a . Gen a -> Gen (m a)) -> Gen a -> Gen (m a)) -- ^ A higher-order generator producing operations using any effects in @m@.
+  -> Gen a                                                              -- ^ A generator for results.
+  -> Gen (m a)                                                          -- ^ A generator producing computations.
 m with = go where
-  go :: forall a . Show a => Gen a -> Gen (With (m a))
-  go a = recursive choice [(liftWith "pure" pure . showing) <$> a] [with go a]
+  go :: forall a . Gen a -> Gen (m a)
+  go a = Gen $ recursive Hedgehog.Gen.choice
+    [ runGen (Gen.label "pure" pure <*> a) ]
+    [ frequency
+      [ (3, runGen (with go a))
+      , (1, runGen (addLabel ">>" (infixL 1 ">>" (>>) <*> go a <*> go a)))
+      ]
+    ]
 
 
-genT :: Gen (T a)
-genT = T <$> integral (linear 0 100)
+genT :: KnownSymbol s => Gen (T s)
+genT = Gen (showing . T <$> integral (linear 0 100))
 
 newtype T a = T { unT :: Integer }
-  deriving (Enum, Eq, Generic, Num, Ord, Real, Vary)
+  deriving (Enum, Eq, Fn.Generic, Num, Ord, Real, Fn.Vary)
 
-instance Arg (T a)
+instance Fn.Arg (T a)
+
+instance S.Semigroup (T a) where
+  T a <> T b = T (a + b)
+
+instance Monoid (T a) where
+  mempty = T 0
+  mappend = (S.<>)
```

Comment:
Defining these instances allows us to avoid using `[]` for `w`, which means much more efficient writer tests.